solver/llbsolver/bridge.go

```diff
@@ -170,6 +172,50 @@ func (b *llbBridge) Solve(ctx context.Context, req frontend.SolveRequest, sid st
 	return
 }
 
+func (b *llbBridge) SaveImage(ctx context.Context, refs map[string]cache.ImmutableRef, metadata map[string][]byte) error {
+	// generate an ID that's consistent for the refs
+	refKeys := []string{}
+	for k := range refs {
+		refKeys = append(refKeys, k)
+	}
+	id := strings.Join(refKeys, "-")
+
+	inp := exporter.Source{
+		Refs:     refs,
+		Metadata: metadata,
+	}
+
+	var exp *ExporterRequest
+	numExporters := 0
+	b.builder.EachValue(context.TODO(), keyEarthlyExporterInstance, func(v interface{}) error {
+		numExporters++
+		exp = v.(*ExporterRequest)
+		return nil
+	})
+	if numExporters != 1 {
+		return fmt.Errorf("SaveImage found %d exporters (should have been 1)", numExporters) // shouldn't happen
+	}
+
+	if e := exp.Exporter; e != nil {
+		err := inBuilderContext(ctx, b.builder, e.Name(), id, func(ctx context.Context, g session.Group) error {
+			sessionIDs := session.AllSessionIDs(g)
+			if len(sessionIDs) == 0 {
+				return fmt.Errorf("group has no session IDs") // shouldnt happen
+			}
+			sessionID := sessionIDs[0]
+			var err error
+			_, err = e.Export(ctx, inp, sessionID)
+			return err
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("SaveImage had no exporter configured")
+	}
```

Comment:
Nit: Slightly easier to read?

(Turned the if-else->error into an if->error)

```suggestion
	e := exp.Exporter
	if e == nil {
		return fmt.Errorf("SaveImage had no exporter configured")
	}
	err := inBuilderContext(ctx, b.builder, e.Name(), id, func(ctx context.Context, g session.Group) error {
		sessionIDs := session.AllSessionIDs(g)
		if len(sessionIDs) == 0 {
			return fmt.Errorf("group has no session IDs") // shouldnt happen
		}
		sessionID := sessionIDs[0]
		var err error
		_, err = e.Export(ctx, inp, sessionID)
		return err
	})
	if err != nil {
		return err
	}
}
```